pkg/reconciler/deployment/controller.go

```diff
@@ -63,21 +66,27 @@ func NewController(
 
 	r.Logger.Info("Setting up event handlers")
 
-	sentinel := impl.EnqueueSentinel(types.NamespacedName{Name: deploymentName, Namespace: namespace})
+	sentinel := impl.EnqueueSentinel(types.NamespacedName{Namespace: namespace, Name: deploymentName})
 	secretInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
 		FilterFunc: controller.FilterWithNameAndNamespace(namespace, secretName),
-		Handler:    handleUpdateOnly(sentinel),
+		Handler:    handler(sentinel),
 	})
 	return impl
 }
 
-func handleUpdateOnly(h func(interface{})) cache.ResourceEventHandler {
+func handler(h func(interface{})) cache.ResourceEventHandler {
 	return cache.ResourceEventHandlerFuncs{
-		AddFunc:    doNothing,
+		// For AddFunc, only enqueue deployment key when envKey is not set.
+		// In such case, the controller pod haven't restarted before.
+		// This helps to avoid infinite loop for restarting controller pod.
+		AddFunc: func(obj interface{}) {
+			if _, ok := os.LookupEnv(envKey); !ok {
+				h(obj)
+			}
+		},
 		UpdateFunc: controller.PassNew(h),
-		DeleteFunc: doNothing,
+		// If secret is deleted, the controller pod will restart, in order to unset the envKey.
+		// This is needed when changing authentication configuration from k8s Secret to Workload Identity.
+		DeleteFunc: h,
```

Comment:
what is the difference between `h` and `controller.PassNew(h)`.

Reply:
The structure of `ResourceEventHandlerFuncs` is 
```
type ResourceEventHandlerFuncs struct {
	AddFunc    func(obj interface{})
	UpdateFunc func(oldObj, newObj interface{})
	DeleteFunc func(obj interface{})
}
```
For UpdateFunc, it is changing the old object (resource before update) to be the new object (resource after update). By using PassNew, only new object is forwarded. Checking the comments under PassNew Function, it is mentioned: PassNew makes it simple to create an UpdateFunc for use with cache.ResourceEventHandlerFuncs that can delegate the same methods as AddFunc/DeleteFunc but passing through only the second argument (which is the "new" object).